test/l4lb-xdp/test.sh

```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+
+set -eux
+
+IMG=${1:-quay.io/cilium/cilium-ci:latest}
+
+###########
+#  SETUP  #
+###########
+
+# bpf_xdp_veth_host is a dummy XDP program which is going to be attached to LB
+# node's veth pair end in the host netns. When bpf_xdp, which is attached in
+# the container netns, forwards a LB request with XDP_TX, the request needs to
+# be picked in the host netns by a NAPI handler. To register the handler, we
+# attach the dummy program.
+clang -O2 -Wall -target bpf -c bpf_xdp_veth_host.c -o bpf_xdp_veth_host.o
+
+# The worker (aka backend node) will receive IPIP packets from the LB node.
+# To decapsulate the packets instead of creating an ipip dev which would
+# complicate network setup, we will attach the following program which
+# terminates the tunnel.
+# The program is taken from the Linux kernel selftests.
+clang -O2 -Wall -target bpf -c test_tc_tunnel.c -o test_tc_tunnel.o
```

Comment:
> these two programs are both hacks that facilitate testing but differ from a typical user environment

Yep.

The `test_tc_tunnel` program is used to terminate the IPIP tunnel on a backend node. Alternatively, we could have created an IPIP tunnel dev. But that approach would be more cumbersome, as we would need to allocate an IP for it from a different subnet, make that subnet routable, etc. I think the chosen approach can be even legit for prod envs.

Meanwhile, the `bpf_xdp_veth_host` is a hack to overcome the veth XDP limitation. I doubt that any user would run XDP on veth (maybe `XDP_REDIRECT`, but that's not relevant to the standalone LB case).

> What's the risk that we miss a regression in some way because of that?

I don't see any big risk. The introduced GH Action is testing the standalone LB capabilities in bpf_host on XDP. The program execution does not depend on any XDP implementation details specific to a driver (minus the hack which does almost nothing). We could have chosen the full-blown Vagrant setup with VMs, but I don't see what additional value it could bring except for an increased CI resource usage and a longer test wait time. Catching any regression in an XDP driver code is not in the scope of these tests.

Reply:
Sounds good to me. Thanks for explaining!